applications/HDF5Application/custom_io/hdf5_file.cpp

```diff
@@ -662,272 +622,229 @@ void File::SetEchoLevel(int Level)
+    ssize_t num_open_objects = H5Fget_obj_count(m_file_id, H5F_OBJ_ALL);
+    KRATOS_ERROR_IF(num_open_objects < 0) << "H5Fget_obj_count failed." << std::endl;
+    return num_open_objects;
     KRATOS_CATCH("");
 }
 
-void File::ReadDataSetIndependent(const std::string& rPath, Vector<int>& rData, unsigned StartIndex, unsigned BlockSize)
+hid_t File::GetFileId() const
 {
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
+    return m_file_id;
 }
 
-void File::ReadDataSetIndependent(const std::string& rPath, Vector<double>& rData, unsigned StartIndex, unsigned BlockSize)
+template<class TDataType>
+bool File::HasDataType(const std::string& rPath) const
 {
     KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
-}
 
-void File::ReadDataSetIndependent(const std::string& rPath, Vector<array_1d<double, 3>>& rData, unsigned StartIndex, unsigned BlockSize)
-{
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
-    KRATOS_CATCH("");
-}
+    hid_t dset_id, dtype_id;
+
+    KRATOS_ERROR_IF((dset_id = H5Dopen(GetFileId(), rPath.c_str(), H5P_DEFAULT)) < 0)
+        << "H5Dopen failed." << std::endl;
+    KRATOS_ERROR_IF((dtype_id = H5Dget_type(dset_id)) < 0)
+        << "H5Dget_type failed." << std::endl;
+    H5T_class_t type = H5Tget_class(dtype_id);
+
+    KRATOS_ERROR_IF(type == H5T_NO_CLASS) << "Invalid data type." << std::endl;
+    KRATOS_ERROR_IF(H5Tclose(dtype_id) < 0) << "H5Tclose failed." << std::endl;
+    KRATOS_ERROR_IF(H5Dclose(dset_id) < 0) << "H5Dclose failed." << std::endl;
+
+    if constexpr(std::is_same_v<TDataType, int>) {
+        return (type == H5T_INTEGER);
+    } else if constexpr(std::is_same_v<TDataType, double>) {
+        return (type == H5T_FLOAT);
+    } else {
+        static_assert(!std::is_same_v<TDataType, TDataType>, "Unsupported data type.");
+    }
 
-void File::ReadDataSetIndependent(const std::string& rPath, Matrix<int>& rData, unsigned StartIndex, unsigned BlockSize)
-{
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
     KRATOS_CATCH("");
 }
 
-void File::ReadDataSetIndependent(const std::string& rPath, Matrix<double>& rData, unsigned StartIndex, unsigned BlockSize)
+template<class TDataType>
+void File::GetDataSet(
+    hid_t& rDataSetId,
+    hid_t& rDataSpaceId,
+    const std::vector<hsize_t>& rDims,
+    const std::string& rPath)
 {
-    KRATOS_TRY;
-    KRATOS_ERROR << "Calling the base class method. Please override in the derived class." << std::endl;
+    KRATOS_TRY
+
+    if (!HasPath(rPath)) {
+        CreateNewDataSet(rDataSetId, rDataSpaceId, Internals::GetPrimitiveH5Type<TDataType>(), rDims, rPath);
```

Comment:
why? shouldn't you throw when you want to get something that doesn't exist?

Reply:
Now this method is moved where it is used, since the new implementation requires it to be called from one place only.